sol/Usdx.sol

```diff
@@ -1,48 +1,148 @@
+		usdPriceFeed = newFeed;
 	}
 
+	// Received eth is minted into usdx at the current eth/usd
+	// exchange rate.
+	// Note: msg.sender must be payable to allow redemption of
+	// deposited eth.
 	receive() external payable {
-		(,int256 rate,,,) = ethUsdPriceFeed.latestRoundData();
-		uint256 usd = msg.value.mul(rate.toUint256()).div(10**FEED_DECS); //latestRoundData uses 8 decimals
-		_mint(_msgSender(), usd);
+		(,int256 rate,,,) = usdPriceFeed.latestRoundData();
+		uint256 toMint = weiToUSDX(msg.value, rate);
+		account storage acct = accounts[msg.sender];
+		acct.locked += msg.value;
+		acct.mint += toMint;
+		_mint(msg.sender, toMint);
+	}
+
+	// Redeems usdx back into eth.  This method can only be called by
+	// senders who have previously directly sent eth into this
+	// contract.  The amount of eth redeemed is capped to the amount
+	// of eth the sender has previously sent.  The amount of eth
+	// redeemed is unrelated to the current ETH/USD price; it's based
+	// purely on the ratio of previously sent eth and previously
+	// minted usdx (ie: if 1 eth was previously received by this
+	// contract to mint 1000usdx, 500 usdx is able to be redeemed for
+	// .5 eth regardless of whether usd/eth price has decreased to
+	// 800usd/eth).  If _amount is 0, msg.sender's full usdx balance
+	// will be used to redeem eth.  Otherwise, _amount is the maximum
+	// amount of usdx used to redeem into eth.
+	function redeem(uint256 _amount) public returns (uint256) {
+		account storage acct = accounts[msg.sender];
+		require(acct.locked > 0, "nothing to redeem");
+		if (_amount == 0) {
+			_amount = acct.mint;
+		} else {
+			_amount = min(_amount, acct.mint);
+		}
+
+		_amount = min(_amount, balanceOf(msg.sender));
+		require(_amount > 0, "no usdx balance");
+		_burn(msg.sender, _amount);
+
+		uint256 unlock = acct.locked.mul(_amount).div(acct.mint);
+		if (_amount == acct.mint) {
+			delete accounts[msg.sender];
+		} else {
+			acct.mint -= _amount;
+			acct.locked -= unlock;
+		}
+		payable(msg.sender).transfer(unlock);
+		return _amount;
+	}
+
+	// Appreciation occurs when previously locked eth appreciates in
+	// usd price. The amount of appreciation can be collected as new
+	// usdx, up to _limit.  A _limit of 0 collects all available
+	// appreciation. To redeem the locked eth, both the principle usdx
+	// mint and any collected appreciation must be returned.
+	function collectAppreciation(uint256 _limit) public returns (uint256) {
+		account storage acct = accounts[msg.sender];
+		(bool ok, uint256 appr) = acctAppreciation(acct);
+		if (!ok) {
+			return 0;
+		}
+		if (_limit > 0) {
+			appr = min(_limit, appr);
+		}
+		acct.mint += appr;
+		_mint(msg.sender, appr);
+		return appr;
+	}
+
+	// transferAcct will transfer the sender's locked eth to _to.  It
+	// does not transfer any usdx balance.  _to must not already have
+	// a usdx account.
+	function transferAcct(address _to) public {
+		require(accounts[_to].locked == 0);
```

Comment:
Do you think `require(accounts[_to].mint == 0)` should also be added? Just thinking that transferring this way might erase any existing USDX that the `_to` address has

Reply:
User essentially has 2 "accounts" that are managed:
1. USDX account (`mapping (address => account) public accounts;`).  This keeps track of usdx that the user has minted.
2. USDX ERC20 account.  This is managed in the openzeppelin/ERC20 contract, and keeps track of the current USDX token balance.

In this function, we're transferring the first account, but not the 2nd (this is called out in the function documentation `It does not transfer any usdx balance.`).  It assumes that if locked == 0, than mint == 0, because locked and mint will increase proportionally together, and decrease proportionally together, but always converge on 0 together.  If you think this assumption breaks anywhere, than we'll need to go over it because it would fundamentally break the solvency of the contract.

Reply:
Ah ok, I think I was just a little confused with the ERC20 account. Makes sense now, thanks